### PR TITLE
Loki: Add autocomplete updates for improved suggestions

### PR DESCRIPTION
**Pull request by ivanahuckova**
This PR improves autocomplete suggestion in Loki in following ways:
- It shows detected parser if user types pipeline and then does space. Previously query failed because of trailing pipeline. Also I have added logic to show `detected` next to parser only if there is no parser in the query.
Before: 

https://user-images.githubusercontent.com/30407135/225038786-7af9deaa-3af5-4582-9343-5b42bf7c443b.mov

After:

https://user-images.githubusercontent.com/30407135/225039561-5d70e06f-0a74-4d86-97dc-ace7bcd12fe7.mov

- Removes showing of `line filters` after pipeline operations (after parser is used), as this is not valid logQl
Before: 

https://user-images.githubusercontent.com/30407135/225038832-17d70501-fa82-40cf-bcbd-556a5a9fa43a.mov

After:

https://user-images.githubusercontent.com/30407135/225039761-82b37d7b-f6bd-428f-b113-5012719dc9bd.mov

- Reorders pipeline operations so that line formatting and label formatting is not hidden on the bottom 
Before: 

https://user-images.githubusercontent.com/30407135/225038985-1d7a9875-0276-4c71-b987-d05b32220460.mov

After:

https://user-images.githubusercontent.com/30407135/225040861-88877564-ce33-49e7-b280-278cb69340a2.mov

- Removes `unwrap label` as this is not correct suggestion. We already do correct suggestions (where we check for label values) after user selects `unwrap` (in `AFTER_UNWRAP` situation). I have changed this to show possible labels for label filter and in the future we can add also label values suggestion. Moreover, `unwrap` is valid only in metric queries, so this could have been confusing for user.
Before: 

https://user-images.githubusercontent.com/30407135/225039042-e14e4ee7-24ab-485d-b217-3eb8f4325595.mov

After:

https://user-images.githubusercontent.com/30407135/225041487-dfacfa1b-28f4-45c7-9d90-acdc2b353c64.mov

I have added tests for all of these.

### CODE REVIEW

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 1]
```diff
@@ -304,6 +305,30 @@ export function getStreamSelectorsFromQuery(query: string): string[] {
   return labelMatchers;
 }
 
+export function removeTrailingPipelines(query: string): string {
+  const trailingPipelinePositions: Position[] = [];
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:309]: If I understand this correctly, this is not just trailing pipelines but all pipelines in the query, right?

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 2]
```diff
@@ -304,6 +305,30 @@ export function getStreamSelectorsFromQuery(query: string): string[] {
   return labelMatchers;
 }
 
+export function removeTrailingPipelines(query: string): string {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:308]: Would you mind adding a few unit tests for this function in `queryUtils.test`? Also, could this be doing something similar as using https://lodash.com/docs/4.17.15#trimEnd ?

Comment (matyax) [reply public/app/plugins/datasource/loki/queryUtils.ts:308]: Also I'm not sure if this is the desired behavior, but when there's not trailing pipe it seems to return an empty query.

imagen

**Review by matyax — COMMENTED** [review]
Looks great! Left a few comments.

File: public/app/plugins/datasource/loki/queryUtils.ts (hunk 2 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/queryUtils.ts:308]: Good suggestion! I'll use https://lodash.com/docs/4.17.15#trimEnd to make it more simple.

File: public/app/plugins/datasource/loki/queryUtils.ts [hunk 3]
```diff
@@ -25,11 +25,12 @@ import {
   MetricExpr,
   Matcher,
   Identifier,
+  PipelineStage,
 } from '@grafana/lezer-logql';
 
 import { ErrorId } from '../prometheus/querybuilder/shared/parsingUtils';
 
-import { getStreamSelectorPositions } from './modifyQuery';
+import { getStreamSelectorPositions, Position } from './modifyQuery';
```
Comment (matyax) [inline public/app/plugins/datasource/loki/queryUtils.ts:33]: Unused import.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts [hunk 4]
```diff
@@ -216,31 +223,30 @@ async function getParserCompletions(
...
 ): Promise<Completion[]> {
-  const { extractedLabelKeys, hasJSON, hasLogfmt } = await dataProvider.getParserAndLabelKeys(logQuery);
-
-  const prefix = `${hasSpace ? '' : ' '}${afterPipe ? '' : '| '}`;
-  const completions: Completion[] = await getParserCompletions(prefix, hasJSON, hasLogfmt, extractedLabelKeys);
+  let query = logQuery;
+  if (afterPipe) {
+    // If we have space after pipeline, we need to remove it as well
+    const trailingPipeline = hasSpace ? '| ' : '|';
+    query = trimEnd(logQuery, trailingPipeline);
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:236]: I would default to this, which works in both cases:
```suggestion
    query = trimEnd(logQuery, '| ');
```

**Review by matyax — APPROVED** [review]
Nice work!

**Pull request merged**